src/qinfer/smc.py

```diff
@@ -542,88 +542,115 @@ def resample(self):
+        :param expparams: The experiments at which to compute the risk.
         :type expparams: :class:`~numpy.ndarray` of dtype given by the current
             model's :attr:`~qinfer.abstract_model.Simulatable.expparams_dtype` property,
             and of shape ``(1,)``
 
-        :return float: The Bayes risk for the current posterior distribution
-            of the hypothetical experiment ``expparams``.
+        :return np.ndarray: The Bayes risk for the current posterior distribution
+            at each hypothetical experiment in ``expparams``, therefore 
+            has shape ``(expparams.size,)``
         """
-        # This subroutine computes the bayes risk for a hypothetical experiment
-        # defined by expparams.
-
-        # Assume expparams is a single experiment
-
-        # expparams =
-        # Q = np array(Nmodelparams), which contains the diagonal part of the
-        #     rescaling matrix.  Non-diagonal could also be considered, but
-        #     for the moment this is not implemented.
-        nout = self.model.n_outcomes(expparams) # This is a vector so this won't work
-        w, N = self.hypothetical_update(np.arange(nout), expparams, return_normalization=True)
-        w = w[:, 0, :] # Fix w.shape == (n_outcomes, n_particles).
-        N = N[:, :, 0] # Fix L.shape == (n_outcomes, n_particles).
-
-        xs = self.particle_locations.transpose([1, 0]) # shape (n_mp, n_particles).
-
-        # In the following, we will use the subscript convention that
-        # "o" refers to an outcome, "p" to a particle, and
-        # "i" to a model parameter.
-        # Thus, mu[o,i] is the sum over all particles of w[o,p] * x[i,p].
-
-        mu = np.transpose(np.tensordot(w,xs,axes=(1,1)))
-        var = (
-            # This sum is a reduction over the particle index and thus
-            # represents an expectation value over the diagonal of the
-            # outer product $x . x^T$.
-
-            np.transpose(np.tensordot(w,xs**2,axes=(1,1)))
-            # We finish by subracting from the above expectation value
-            # the diagonal of the outer product $mu . mu^T$.
-            - mu**2).T
 
+        # for models whose outcome number changes with experiment, we 
+        # take the easy way out and for-loop over experiments
+        n_eps = expparams.size
+        if n_eps > 1 and not self.model.is_n_outcomes_constant:
+            risk = np.empty(n_eps)
+            for idx in range(n_eps):
+                risk[idx] = self.bayes_risk(expparams[idx, np.newaxis])
+            return risk
+
+        # but if we make it here, the following should be a single number
+        n_out = self.model.n_outcomes(expparams)
+        
+        # compute the hypothetical weights, likelihoods and normalizations for
+        # every possible outcome and expparam
+        # the likelihood over outcomes should sum to 1, so don't compute for last outcome
+        w_hyp, L, N = self.hypothetical_update(
```

Comment:
I really like this optimization, it should help a lot with models that have a small number of outcomes. Since the same optimization appears in ``expected_information_gain``, perhaps it's worth abstracting / refactoring it out?